src/Core/Framework/Adapter/Cache/InvalidatorStorage/MySQLInvalidatorStorage.php

```diff
@@ -0,0 +1,104 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Framework\Adapter\Cache\InvalidatorStorage;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\TransactionIsolationLevel;
+use Shopware\Core\Framework\DataAbstractionLayer\Doctrine\MultiInsertQueryQueue;
+use Shopware\Core\Framework\DataAbstractionLayer\Doctrine\RetryableQuery;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Uuid\Uuid;
+
+/**
+ * Whilst this adapter attempts to work around various locking issues, it is not foolproof, so if you encounter deadlock due to load, use the `\Shopware\Core\Framework\Adapter\Cache\InvalidatorStorage\RedisInvalidatorStorage` adapter instead.
+ *
+ * @codeCoverageIgnore @see \Shopware\Tests\Integration\Core\Framework\Adapter\Cache\InvalidatorStorage\MySQLInvalidatorStorageTest
+ */
+#[Package('core')]
+class MySQLInvalidatorStorage extends AbstractInvalidatorStorage
+{
+    private const TABLE_NAME = 'invalidation_tags';
+
+    private readonly \Closure $debug;
+
+    public function __construct(private readonly Connection $connection, ?\Closure $debug = null)
+    {
+        $this->debug = $debug ?? (fn () => null)(...);
+    }
+
+    public function store(array $tags): void
+    {
+        if (empty($tags)) {
+            return;
+        }
+
+        $insertQueue = new MultiInsertQueryQueue($this->connection, chunkSize: 1000, ignoreErrors: true);
+        $insertQueue->addInserts(
+            self::TABLE_NAME,
+            array_map(
+                fn (string $tag) => ['id' => Uuid::randomBytes(), 'tag' => $tag],
+                array_values($tags)
+            )
+        );
+
+        // we execute in read committed isolation so that row gap locks are not applied when inserting
+        // this helps us prevent locks when trying to insert duplicate tags.
+        $this->readCommittedIsolation(fn () => $insertQueue->execute());
+    }
+
+    /**
+     * We attempt to read and lock all rows in the table. This works fine for subsequent executions, e.g. calling this method multiple times synchronously
+     * however, if running in parallel, it can still be that the database returns similar result sets when it did not get a chance to lock the rows. Then, when trying to
+     * delete, locks can be encountered if two processes attempt to delete the same row.
+     */
+    public function loadAndDelete(): array
+    {
+        return $this->readCommittedIsolation(
```

Comment:
@keulinho I think we should wrap the whole thing in a try/catch and return an empty array if a deadlock occurs? We could also log it. wdyt ?

Reply:
Makes sense, should we log at level warning?